agent/command.go

```diff
@@ -165,15 +166,20 @@ func (a *Agent) runCommand(ctx context.Context, tc *taskContext, logger client.L
 	}
 	defer func() {
 		if !tc.unsetFunctionVarsDisabled || tc.project.UnsetFunctionVars {
-			// This defer ensures that the function vars do not persist in the expansions after the function is over.
-			for key, functionValue := range commandInfo.Vars {
-				currentValue := tc.taskConfig.Expansions.Get(key)
-				if currentValue != functionValue {
-					// If a command in the func updates the expansion value, don't reset it.
-					delete(prevExp, key)
+			// This defer ensures that the function vars do not persist in the expansions after the function is over
+			// unless they were updated using expansions.update
+			if cmd.Name() == "expansions.update" {
+				updatedExpansions := tc.taskConfig.DynamicExpansions.Map()
+				for k := range updatedExpansions {
+					if _, ok := commandInfo.Vars[k]; ok {
+						// If expansions.update updated this key, don't reset it
+						delete(prevExp, k)
+					}
 				}
 			}
 			tc.taskConfig.Expansions.Update(prevExp)
+			tc.taskConfig.DynamicExpansions = util.EmptyExpansion()
```

Comment:
I'm pretty sure you can just call `util.NewExpansions(map[string]string{})` rather than make a brand new function for a specific case of it.